### DISCUSSION TITLE
custom data on schema

### DISCUSSION THREAD

**Discussion by leolmi** [colinhacks/zod#2046] [Ideas]
it could be very useful to have the possibility to add customizable attributes in the definitions of a schema:
````js
const Foo = z.object({
  myProperty: z.string({
    description: 'my property description',
    customData: {
      // my custom data
    }
  })
});
````
or something like
````js
const Foo = z.object({
  myProperty: z.string().customData({
      // my custom data
  })
});
````
so they would be present in the schema
````js
export interface JSONSchema7 {
  ...
  customData?: any;
  ...
}
````
and they would be accessible from the serialized schema:
````js
const schema = zodToJsonSchema(Foo, {$refStrategy: 'none'});

const myProperty = (<JSONSchema7>schema).properties['myProperty'];

console.log('custom data from schema', myProperty.customData); 
````

**Comment by JacobWeisenburger** [comment]
Can you not just do something like this?

```ts
const Foo = {
    customData: {
        // my custom data
    },
    schema: z.object( {
        myProperty: z.string( {
            description: 'my property description',
        } )
    } )
}
```

**Reply by leolmi** [reply]
it's not quite the same thing: in this way I should maintain two parallel models.
But I just saw one of the last commits you made: [https://github.com/colinhacks/zod/pull/2042](https://github.com/colinhacks/zod/pull/2042), which is about "_arbitrary schema annotations_", and this probably solves my request.

**Reply by stephenh** [reply]
Oh nice! This would be great; I have a library that currently uses a non-zod/custom API that looks like:

```
    const a = createObjectState<AuthorInput>(
      {
        books: {
          type: "list",
          config: {
            title: { type: "value", rules: [required] },
            delete: { type: "value", isDeleteKey: true },
          },
        },
      },
      {},
    );
```

Which could be like `{ delete: z.boolean() }` or what not, but I need a spot for `isDeleteKey` to go, b/c that's special to my application.

This API is definitely doable:

```ts
const isPii = Symbol("isPii");
const annotatedString = z
  .string()
  .annotate(isPii, true);
annotatedString.meta[isPii]; // true
```

But @nicknotfun FWIW if there was a way to add my own fluent methods, like:

```
const annotatedString = s.string().isPii();
```

That would be amazing.

Maybe via TypeScript's declaration merging? Like on the Tanstack router...

https://tanstack.com/router/v1/docs/guide/type-safety#exported-hooks-components-and-utilities

Like let me do some sort of:

```
declare module 'zod' {
  interface ... {
    isPii(): boolean
  }
}
import { registerAnnotation } from "zod";
registerAnnotation("isPii");
```

Dunno, something like that. Thanks!

**Reply by nicknotfun** [reply]
I do like the idea of a fluent API, but couldn't figure an easy way to do it. I'd just love something like this at-minimum [we're forking zod to use this for our own purposes]. I don't think my annotation data precludes a fluent interface

**Reply by stephenh** [reply]
Ah yep, @nicknotfun agreed that your annotation approach doesn't preclude the fluent interface, and would probably be a great place for some eventual fluent API to store it's data in...

**Reply by nicknotfun** [reply]
Btw @leolmi - the commit you refer to is my feature request - I am not a maintainer so unsure when it'd get integrated.

**Reply by leolmi** [reply]
Ok thanks @nicknotfun , I hope they consider your implementation soon
